### PR TITLE
#53 support babashka use httpkit + api updates fro vault 1.7.1 + deps.edn

### LINKED ISSUE #53: Support babashka
Hi, 

We are using babashka tasks to deploy. 
We store secrets in vault and would like to get secrets using babashka tasks as well. 

Would you consider supporting babashka (implementing or accepting PR's ) ? 

From the look at the library, the dependencies are quite common and should work. 
`clojure.tools.logging` does not seem to work. 
Babashka does not have built-in logging but this could be used: https://gist.github.com/borkdude/c97da85da67c7bcc5671765aef5a89ad

Babashka can interpret clojure code using sci and supports quite a few clojure libraries https://github.com/babashka/babashka/blob/master/doc/projects.md .

This could be achieved if the bits that are not supported are re-implemented with bits that are supported by babashka.

Bellow is a ample babashka script to help with development.
bb.edn
```
;; Type bb tasks to see all tasks
;; Type bb <task-name> or bb run <task-name> to run a task
{:min-bb-version "0.4.0"
 :paths ["script"]
 :deps {amperity/vault-clj {:mvn/version "1.0.3"}}
 :tasks {:requires ([vault.core :as vault]
                    [vault.client.http])
         :init     (do 
                       (def client (vault/new-client "https://vault.do.drevidence.com:8200")))
         ;; Helpers
         vault-get-dev {:doc "Get dev secrets from vault - requires vault cli."
                            :task (do 
                                    (client))}
}}

```

```
bb vault-get-dev
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Could not resolve symbol: clojure.lang.LockingTransaction/isRunning
Location: clojure/tools/logging.clj:59:15
Phase:    analysis

----- Context ------------------------------------------------------------------
55:   [logger level throwable message]
56:   (if (case *force*
57:         :agent  true
58:         :direct false
59:         (and (clojure.lang.LockingTransaction/isRunning)
                  ^--- Could not resolve symbol: clojure.lang.LockingTransaction/isRunning
60:              (*tx-agent-levels* level)))
61:     (send-off *logging-agent*
62:       (fn [_#] (impl/write! logger level throwable message)))
63:     (impl/write! logger level throwable message))
64:   nil)
```

### PR DESCRIPTION
**Pull request by ieugen**
This is a rough PR - mainly to visualize the amount of changes. 

* I migrated from clj-http to httpkit (API is mostly compatible) - minimal changes
* Added deps.edn for easier consumption of library via git url or local url
* Added bb.edn with sample script to test secret extraction from vault
* Changed clean-body function to make it work with vault 1.7.x API
* Replaced commons hex with clojure equivalent

Fixes #53 

Ideas for improvement:
* Add integration tests using https://www.testcontainers.org/  - start vault in a container and run API tests against it

### CODE REVIEW

**Review by pyncc — CHANGES_REQUESTED** [review]
Nice, thanks for taking the initiative to solve the issue you're facing.

This seems like a reasonable approach, just had a few questions and requests about some of these changes.

Also, congratulations github says this is your first time contributing! 🎉

File: src/vault/client/api_util.clj [hunk 1]
```diff
@@ -123,23 +130,22 @@
...
+  ;; Check API path.
   (when-not (and (string? path) (not (str/blank? path)))
     (throw (IllegalArgumentException.
              (str "API path must be a non-empty string, got: "
                   (pr-str path)))))
-  ; Check client authentication.
+  ;; Check client authentication.
   (when-not (some-> client :auth deref :client-token)
-    (throw (IllegalStateException.
+    (throw (RuntimeException.
```
[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/client/api_util.clj:140]: IllegalStateException was missing from bb https://github.com/babashka/babashka/pull/941 .

File: src/vault/lease.clj [hunk 2]
```diff
@@ -199,8 +200,8 @@
 
 (defn ^:no-doc maintain-leases!
   [client window]
-  (log/trace "Checking for renewable leases...")
-  ; Check auth token for renewal.
+  (log/debug "Checking for renewable leases...")
```
[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/lease.clj:203]: It was also missing from bb and was added in the mean time. 
My focus was to make it work.

File: src/vault/timer.clj [hunk 3]
```diff
@@ -23,7 +23,8 @@
             (catch Exception ex
               (log/error ex "Exception while running timer handler!")))
           (recur)))
-      (catch InterruptedException _
+      (catch Exception e
```
[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/timer.clj:26]: It's missing from babashka. 
I'll see if it makes sense to add it or look for an alternative ( compile conditionals ?!) 
cc @borkdude

File: project.clj [hunk 4]
```diff
@@ -5,28 +5,32 @@
             :url "http://www.apache.org/licenses/LICENSE-2.0"}
 
   :deploy-branches ["master"]
-  :pedantic? :abort
+  ;; :pedantic? :abort
```
[the remark this answers is not in the document]
Comment (ieugen) [reply project.clj:8]: it was failing locally for me. Don't know why.
Now it works.
Restored.

File: src/vault/client/api_util.clj [hunk 5]
```diff
@@ -154,5 +160,13 @@
       (:http-opts client)
       {:headers {"X-Vault-Token" wrap-token}
        :content-type :json
-       :accept :json
-       :as :json})))
+       :accept :json})))
+
+
+(comment
```
[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/client/api_util.clj:166]: done.

File: src/vault/client/api_util.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/client/api_util.clj:140]: It's added in master but not yet released so I guess this is done.

File: dev/user.clj [hunk 6]
```diff
@@ -34,3 +38,22 @@
     (when vault-token
       (vault/authenticate! vault-client :token vault-token))
     (alter-var-root #'client (constantly (component/start vault-client)))))
+
```
[the remark this answers is not in the document]
Comment (ieugen) [reply dev/user.clj:41]: It's not necessary to keep them but they where very helpful in debugging.
I think they can help others (new contributors esp.) as well.
I would like to keep them if it's all the same, but don't feel strong about this.

File: project.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (ieugen) [reply project.clj]: It seems to work only with fully qualified name, even with the patch java.lang.IllegalStateException .
This does not make a difference. 
I've asked for help to see why it is like this though.

File: src/vault/timer.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (borkdude) [reply src/vault/timer.clj:26]: Yes, we can add it.

[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/timer.clj:26]: There is https://github.com/babashka/babashka/pull/943 .

[the remark this answers is not in the document]
Comment (ieugen) [reply src/vault/timer.clj]: Using Interrupted exception, tested with local bb build.

File: dev/user.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (ieugen) [reply dev/user.clj:41]: Please reopen if you want me to remove them,

**Comment by ieugen** [comment]
I've implemented the requested changes. 
It should work with bb from master right now.
I tested only read-secrets, but the others should work as well.

**Comment by borkdude** [comment]
@ieugen We can add the unit tests of this library to babashka CI. I tend to do this for libraries that are compatible with bb. See test-resources/lib_tests and script/run_lib_tests.

**Comment by ieugen** [comment]
Thank you @pyncc . 
I'll see if I can write some real integration tests via testcontainers. 

I believe it's better to test against a real, in memory vault instance.

**Comment by pyncc** [comment]
Release to clojars @ 1.0.4

https://github.com/amperity/vault-clj/releases/tag/1.0.4

> I'll see if I can write some real integration tests via testcontainers.

That sounds good to me, please keep us updated!

**Pull request merged**